### ISSUE TITLE
Parameter linting

### ISSUE THREAD

**Issue by dberenbaum** [iterative/dvc#9500]
Request from @JulianWgs: projects often have complex parameter setups, and it's easy to forget something in params.yaml, dvc.yaml, or the user code. Can we have a way to check for parameters defined in one of those places but not in another?

* Are all used parameters defined?
* Are all defined parameters used?

Some examples:
* Parameter issue causes a downstream stage to fail after several stages have run (hours into the project)
* Stage code uses parameters that were not defined as parameters for that stage

cc @efiop @shcheklein

**Comment by dberenbaum** [comment]
> * Are all used parameters defined?

If a parameter is defined in `dvc.yaml` but missing from `params.yaml`, it should show as deleted in `dvc status` like this:

```
$ dvc status
train:
        changed deps:
                params.yaml:
                        deleted:            train.lr
```

Maybe we can improve the output of this to make it more clear what it means and what to do about it?

> * Are all defined parameters used?

AFAIK there's no way to know today whether there are parameters defined in `params.yaml` but not used in `dvc.yaml`. We could add some kind of output to `dvc status` (again, possibly with a flag) that could list unused/undefined parameters from `params.yaml` or any other loaded parameters files.

> * Parameter issue causes a downstream stage to fail after several stages have run (hours into the project)

We could check all stages eagerly when running the pipeline (possibly with a flag) and fail if parameters are missing.



> * Stage code uses parameters that were not defined as parameters for that stage

There are a few options here but none are ideal:

1. Use `dvc.api.params_show(stages="train")`, but that requires users to add `dvc.api` calls to their code.
2. Modify the parameters file during stage execution to drop undefined parameters, but modifying the users' files like this is something we would rather avoid as much as possible.
3. Parse the user code to find what parameters it expects, but this will be hard to do across languages and expands the scope even if we only started with python.

We could consider this item out of scope for now if it unblocks us.

@JulianWgs @efiop Please add your thoughts.

**Comment by efiop** [comment]
> Maybe we can improve the output of this to make it more clear what it means and what to do about it?

Yeah, not sure if this is the best thing to do to shove it into `dvc status`, since it is already pretty bad. Obviously a late warning when we are trying to use params is also not the best idea. Maybe `dvc params show`?
